### PR TITLE
Dark/Light theming

### PR DESCRIPTION
**Pull request by nickbutcher**
Introduce light/dark theme support.

plaid_dark_theme_toggle

- Add a toggle to the main toolbar (defaulted to follow system).
- Update main theme to inherit from `Theme.MaterialComponents.DayNight`.
- Update all widgets to use `MaterialComponents` styles/TextAppearances where possible.
- Update all color references to use theme attributes (`?attr/foo`).
- Provide alternate darker palette in dark theme.
- Darken images in main feed in dark theme.

Some more caps:
1-grid-light
2-grid-dark
3-dn-light
4-dn-dark
5-drib-light
6-drib-dark
7-login-light
8-login-dark
9-about-light
10-about-dark

### CODE REVIEW

File: app/src/main/java/io/plaidapp/ui/HomeActivity.java [hunk 1]
```diff
@@ -94,7 +97,7 @@
 
 import static io.plaidapp.dagger.Injector.inject;
 
-public class HomeActivity extends FragmentActivity {
+public class HomeActivity extends AppCompatActivity {
```
Comment (nickbutcher) [inline app/src/main/java/io/plaidapp/ui/HomeActivity.java]: Not sure how we'd managed so long without this?

File: app/src/main/java/io/plaidapp/ui/HomeActivity.java [hunk 2]
```diff
@@ -343,10 +347,25 @@ public void onLayoutChange(View v, int l, int t, int r, int b,
...
+            final CheckBox toggle = (CheckBox) actionView;
+            toggle.setButtonDrawable(R.drawable.asl_theme);
+            toggle.setChecked(ColorUtils.isDarkTheme(this));
+            toggle.jumpDrawablesToCurrentState();
+            toggle.setOnCheckedChangeListener((buttonView, isChecked) -> {
+                // delay to allow the toggle anim to run
+                toggle.postDelayed(() -> {
+                    AppCompatDelegate.setDefaultNightMode(isChecked ?
+                            AppCompatDelegate.MODE_NIGHT_YES : AppCompatDelegate.MODE_NIGHT_NO);
+                    getDelegate().applyDayNight();
```
Comment (nickbutcher) [inline app/src/main/java/io/plaidapp/ui/HomeActivity.java]: Won't be needed once this CL is released: https://android-review.googlesource.com/c/platform/frameworks/support/+/933658

File: app/src/main/java/io/plaidapp/ui/HomeActivity.java [hunk 3]
```diff
@@ -343,10 +347,25 @@ public void onLayoutChange(View v, int l, int t, int r, int b,
...
-    @Override
-    public boolean onCreateOptionsMenu(Menu menu) {
-        getMenuInflater().inflate(R.menu.main, menu);
-        return true;
+    private void setupToolbar() {
+        toolbar.inflateMenu(R.menu.main);
+        final MenuItem toggleTheme = toolbar.getMenu().findItem(R.id.menu_theme);
+        final View actionView = toggleTheme.getActionView();
+        if (actionView instanceof CheckBox) {
+            final CheckBox toggle = (CheckBox) actionView;
```
Comment (nickbutcher) [inline app/src/main/java/io/plaidapp/ui/HomeActivity.java]: This is something of a UX experiment: instead of offering the usual 3 options (Light, Dark, System) instead we're trying always defaulting (on App start) to system then offering just a light/dark toggle. This change isn't persisted past the app's process so if it's restarted it will revert to system setting.

Again this is an experiment and might change in the future.

**Comment by nickbutcher** [comment]
@chrisbanes Would love a review if possible?

File: core/src/main/res/values/colors.xml [hunk 4]
```diff
@@ -71,20 +67,20 @@
...
     <color name="designer_news_link_highlight">#b303a9f4</color>
     <color name="designer_news_story_comment_background">@android:color/transparent</color>
 
     <color name="designer_news_super_dark">#000133</color>
-    <color name="designer_news_quote_line">#ffeceef1</color>
-    <color name="designer_news_quote">#73000000</color>
 
     <!-- product hunt-->
-    <color name="product_hunt">#ffff9800</color>
+    <color name="product_hunt">#ffff9800</color> <!--orange 500 -->-->
```
Comment (chrisbanes) [inline core/src/main/res/values/colors.xml]: Double `-->`

File: designernews/src/main/java/io/plaidapp/designernews/ui/story/StoryActivity.java [hunk 5]
```diff
@@ -89,9 +88,7 @@
 import java.util.List;
 
 import static com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions.withCrossFade;
-import static io.plaidapp.core.util.AnimUtils.getFastOutLinearInInterpolator;
-import static io.plaidapp.core.util.AnimUtils.getFastOutSlowInInterpolator;
-import static io.plaidapp.core.util.AnimUtils.getLinearOutSlowInInterpolator;
+import static io.plaidapp.core.util.AnimUtils.*;
```
Comment (chrisbanes) [inline designernews/src/main/java/io/plaidapp/designernews/ui/story/StoryActivity.java]: Wildcards are usually avoided?

File: designernews/src/main/java/io/plaidapp/ui/drawable/ThreadedCommentDrawable.kt [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            canvas.drawLine(left.toFloat(), 0f, left.toFloat(), bounds.bottom.toFloat(), paint)
+        }
+    }
+
+    override fun getIntrinsicWidth(): Int {
+        return threads * threadWidth + (threads - 1) * gap
+    }
+
+    override fun setAlpha(i: Int) {
+        paint.alpha = i
```
Comment (chrisbanes) [inline designernews/src/main/java/io/plaidapp/ui/drawable/ThreadedCommentDrawable.kt]: Need to invalidateSelf()

File: designernews/src/main/java/io/plaidapp/ui/drawable/ThreadedCommentDrawable.kt [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    override fun setColorFilter(colorFilter: ColorFilter?) {
+        paint.colorFilter = colorFilter
```
Comment (chrisbanes) [inline designernews/src/main/java/io/plaidapp/ui/drawable/ThreadedCommentDrawable.kt]: Need to invalidateSelf()

File: build.gradle [hunk 8]
```diff
@@ -21,29 +21,30 @@ buildscript { scriptHandler ->
...
+            'appcompat'          : '1.1.0-alpha03',
             'androidx'           : '1.0.0',
             'androidxCollection' : '1.0.0',
             'androidxCoreRuntime': '2.0.1-alpha01',
             'androidxArch'       : '2.0.0',
             'constraintLayout'   : '2.0.0-alpha2',
             'coreKtx'            : '1.0.0',
             'coroutines'         : '1.1.1',
-            'crashlytics'        : '2.9.8',
+            'crashlytics'        : '2.9.9',
```
Comment (florina-muntenescu) [inline build.gradle]: are all of these version increases mandatory for dark/light theming? If not, i suggest moving those to another PR, so you can keep this one focused to the theming.

File: core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt [hunk 9]
```diff
@@ -70,7 +70,8 @@ class FeedAdapter(
     // we need to hold on to an activity ref for the shared element transitions :/
     private val host: Activity,
     private val columns: Int,
-    private val pocketIsInstalled: Boolean
+    private val pocketIsInstalled: Boolean,
+    private val isDarkTheme: Boolean
```
[the remark this answers is not in the document]
Comment (florina-muntenescu) [reply core/src/main/java/io/plaidapp/core/feed/FeedAdapter.kt]: it's used, but never set to true.

File: core/src/main/java/io/plaidapp/core/util/ColorUtils.java [hunk 10]
```diff
@@ -50,16 +51,20 @@ private ColorUtils() { }
     /**
      * Set the alpha component of {@code color} to be {@code alpha}.
      */
-    public static @CheckResult @ColorInt int modifyAlpha(@ColorInt int color,
-                                                         @IntRange(from = 0, to = 255) int alpha) {
+    public static @CheckResult
```
Comment (florina-muntenescu) [inline core/src/main/java/io/plaidapp/core/util/ColorUtils.java]: given that you formatted the file, could you also move the annotations before `public static`?

Comment (florina-muntenescu) [reply core/src/main/java/io/plaidapp/core/util/ColorUtils.java]: or, revert the formatting and leave just `isDarkTheme` method, so it's more visible what's new in this file

File: designernews/src/main/java/io/plaidapp/designernews/ui/login/LoginActivity.kt [hunk 11]
```diff
@@ -82,7 +82,7 @@ class LoginActivity : AppCompatActivity() {
             MorphTransform.setup(
                 this,
                 binding.container,
-                ContextCompat.getColor(this, appR.color.background_light),
+                ColorUtils.getThemeColor(this, io.plaidapp.core.R.attr.colorSurface),
```
Comment (florina-muntenescu) [inline designernews/src/main/java/io/plaidapp/designernews/ui/login/LoginActivity.kt]: consider using `coreR`

File: designernews/src/main/java/io/plaidapp/ui/widget/PasswordEntry.java [hunk 12]
```diff
@@ -44,11 +42,11 @@
...
 public class PasswordEntry extends TextInputEditText {
 
-    static final char[] PASSWORD_MASK = { '\u2022' }; // PasswordTransformationMethod#DOT
+    static final char[] PASSWORD_MASK = {'\u2022'}; // PasswordTransformationMethod#DOT
 
-    private boolean passwordMasked = false;
+    private boolean passwordMasked;
     private MaskMorphDrawable maskDrawable;
-    private ColorStateList textColor;
+    ColorStateList textColor;
```
Comment (florina-muntenescu) [inline designernews/src/main/java/io/plaidapp/ui/widget/PasswordEntry.java]: can this stay `private`?

File: dribbble/src/main/java/io/plaidapp/dribbble/ui/shot/ShotActivity.kt [hunk 13]
```diff
@@ -64,8 +64,6 @@ class ShotActivity : AppCompatActivity() {
         R.layout.activity_dribbble_shot
     )
 
-    private var largeAvatarSize: Int = 0
```
Comment (florina-muntenescu) [inline dribbble/src/main/java/io/plaidapp/dribbble/ui/shot/ShotActivity.kt]: these changes are not related to Dark/Light theming. Move them to a different PR

File: designernews/src/main/java/io/plaidapp/ui/widget/PasswordEntry.java [hunk 14]
```diff
@@ -140,9 +127,9 @@ private int getTextLeft() {
         private final long hidePasswordDuration;
         private final Interpolator fastOutSlowIn;
 
-        private CharSequence password;
-        private PasswordCharacter[] characters;
-        private float morphProgress;
+        CharSequence password;
+        PasswordCharacter[] characters;
+        float morphProgress;
```
Comment (florina-muntenescu) [inline designernews/src/main/java/io/plaidapp/ui/widget/PasswordEntry.java]: these don't seem to be used anywhere else, can they stay private?

File: designernews/src/main/java/io/plaidapp/ui/widget/PasswordEntry.java [hunk 15]
```diff
@@ -230,7 +217,7 @@ public void onAnimationEnd(Animator animation) {
             return anim;
         }
 
-        private void updateBounds() {
+        void updateBounds() {
```
Comment (florina-muntenescu) [inline designernews/src/main/java/io/plaidapp/ui/widget/PasswordEntry.java]: this also doesn't seem to be used anywhere else, can it stay private?

File: dribbble/src/main/res/layout/activity_dribbble_shot.xml [hunk 16]
```diff
@@ -169,7 +169,7 @@
                     app:goneUnless="@{uiModel.shouldHideDescription}"
                     tools:text="Check out this sweet eye candy!" />
 
-                <Button
+                <io.plaidapp.core.ui.widget.BaselineGridTextView
```
Comment (florina-muntenescu) [inline dribbble/src/main/res/layout/activity_dribbble_shot.xml]: is this change related to dark/light theming?

File: app/src/main/java/io/plaidapp/ui/HomeActivity.java (hunk 3 above)
Comment (manuelvicnt) [reply app/src/main/java/io/plaidapp/ui/HomeActivity.java]: Are we tracking any analytics to see how this behaves?

File: app/src/main/java/io/plaidapp/ui/PlaidApplication.kt [hunk 17]
```diff
@@ -27,9 +31,18 @@ import io.plaidapp.core.dagger.DaggerCoreComponent
...
 class PlaidApplication : Application() {
 
+    override fun onCreate() {
+        super.onCreate()
+        val nightMode = if (BuildCompat.isAtLeastQ()) {
+            MODE_NIGHT_FOLLOW_SYSTEM
+        } else {
+            MODE_NIGHT_AUTO_BATTERY
+        }
+        setDefaultNightMode(nightMode)
```
Comment (manuelvicnt) [inline app/src/main/java/io/plaidapp/ui/PlaidApplication.kt]: Do you need `applyDayNight()` here? This won't be necessary in the future but just to keep it consistent

File: app/src/main/res/drawable/dialog_background.xml [hunk 18]
```diff
@@ -19,5 +19,5 @@
     xmlns:android="http://schemas.android.com/apk/res/android"
     android:shape="rectangle">
     <corners android:radius="@dimen/dialog_corners" />
-    <solid android:color="@color/background_light" />
-</shape>
\ No newline at end of file
+    <solid android:color="?attr/colorSurface" />
```
Comment (manuelvicnt) [inline app/src/main/res/drawable/dialog_background.xml]: What's the difference between `?attr/colorSurface` and `@color/color_surface`? Could we merge them? Are they different colours?

File: app/src/main/java/io/plaidapp/ui/PlaidApplication.kt (hunk 17 above)
Comment (chrisbanes) [reply app/src/main/java/io/plaidapp/ui/PlaidApplication.kt]: @manuelvicnt You can't `applyDayNight()` an Application 😄

File: core/src/main/java/io/plaidapp/core/util/ColorUtils.java [hunk 19]
```diff
@@ -69,13 +74,15 @@ private ColorUtils() { }
      * Annoyingly we have to return this Lightness 'enum' rather than a boolean as palette isn't
      * guaranteed to find the most populous color.
      */
-    public static @Lightness int isDark(Palette palette) {
+    public static @Lightness
+    int isDark(Palette palette) {
```
[the remark this answers is not in the document]
Comment (nickbutcher) [reply core/src/main/java/io/plaidapp/core/util/ColorUtils.java]: Reverted.

File: designernews/src/main/java/io/plaidapp/ui/widget/PasswordEntry.java (hunk 12 above)
Comment (nickbutcher) [reply designernews/src/main/java/io/plaidapp/ui/widget/PasswordEntry.java]: It can. This was removing a synthetic accessor but don't feel strongly if we want to optimize for encapsulation.

File: dribbble/src/main/res/layout/activity_dribbble_shot.xml (hunk 16 above)
Comment (nickbutcher) [reply dribbble/src/main/res/layout/activity_dribbble_shot.xml]: Yes.

File: app/src/main/res/drawable/dialog_background.xml (hunk 18 above)
Comment (nickbutcher) [reply app/src/main/res/drawable/dialog_background.xml]: `@color/color_surface` is the value used to set `?attr/colorSurface`. Ideally we'd only need the theme attr but unfortunately there are places that we can't use theme attrs such as in `AnimatedVectorDrawable`s so we need both.

File: app/src/main/res/drawable/ic_comment_add.xml [hunk 20]
```diff
@@ -35,9 +35,9 @@
         <path
             android:name="@string/comment_add"
             android:pathData="@string/path_comment_add"
-            android:strokeColor="@color/comment_line_stroke_color"
+            android:strokeColor="?attr/colorSurface"
             android:strokeLineCap="square"
-            android:strokeWidth="@integer/comment_line_stroke_width" />
+            android:strokeWidth="2" />
```
[the remark this answers is not in the document]
Comment (nickbutcher) [reply app/src/main/res/drawable/ic_comment_add.xml]: Reverted
